### ISSUE TITLE
Sin and Cos Bad Behavior for Large Arguments

### ISSUE THREAD

**Issue by dcwuser** [dotnet/corefx#16486]
For |x| > ~1.0E19, Math.Sin(x) and Math.Cos(x) return x rather than an approximation to Sin(x) or Cos(x). This is documented, but is nonetheless a very bad behavior and should be changed to throw an ArgumentOutOfRangeException or return a NaN.

Calculating trig functions of large arguments is hard for reasons well-known to numerical computing experts and much less well-known to most other programmers. The issue has a long history of sowing confusion. (See https://randomascii.wordpress.com/2014/10/09/intel-underestimates-error-bounds-by-1-3-quintillion/ and some references there.) One way of viewing the problem is that FP doubles have ~16 digits of accuracy, so once you get to values above ~6.0E16, there is a full 2\pi of "uncertainty" in the value, so we don't really know where we are within the period of the sine curve.

The real solution to this problem would be to implement range reduction with a high-precision value of \pi, which would enable us to give 1-ulp accurate values of trig functions for all representable doubles. I will request that solution in another bug. This bug is about what we should do if we are not willing to implement the real solution.

Page 3 of the paper https://www.csee.umbc.edu/~phatak/645/supl/Ng-ArgReduction.pdf has a table that compares the behavior of trig functions for large argument for ~30 different platforms. Some of them implement the real solution and so are able to give the correct answer. Others return what is essentially a random value between -1 and 1, or return 0 for sin and 1 for cos. Almost all of these platforms satisfy sin^2 + cos^2 = 1. Finally, some platforms return an error or a NaN to signal their inability to do the requested calculation. What none of them do is what .NET does: return the argument.

Besides being unique, what .NET does is bad for a number of additional reasons:

(i) One thing we know for certain about sin and cos values is that they must lie between -1 and 1; our behavior violates this rule.

(ii) Another thing we know for certain about sin and cos values is that sin^2 + cos^2 = 1; our behavior violates this rule.

(iii) Typically people don't call Math.Sin and Math.Cos as part of a program that just computes sines and cosines. Instead, the input argument is the result of a complex calculation and the output value feeds into other complex expressions that are supposed to produce the desired numerical result. It is very easy for a program to produce an input in the bad range in a way that is hard for the programmer to foresee. Without any clear error signal, it's also very easy for the program to go on to produce an ultimate numerical result that is not obviously the total bull**** that it in fact is.

(iv) Of course, since we document this, we could argue that the programmer should check input validity before calling Math.Sin or Math.Cos. But if the programmer should always do this, why don't we do it for him? A single floating point comparison is not a high overhead.

There are some applications where speed is of the utmost importance, more important even than validity. Gaming comes to mind. (Actually, pretty much only gaming comes to mind.) But we don't allow such considerations to outweigh the requirement for validity for other functions. No one would suggest that we replace Math.Sqrt with Carmack's famous fast inverse square root trick (https://en.wikipedia.org/wiki/Fast_inverse_square_root); we expect people who prefer a fast-and-usually-approximately-right answer to a slower-but-actually-correct answer to call a dedicated fast-but-approximate method.

**Comment by mikedn** [comment]
> Page 3 of the paper https://www.csee.umbc.edu/~phatak/645/supl/Ng-ArgReduction.pdf has a table that compares the behavior of trig functions for large argument for ~30 different platforms

I'm not sure how that's relevant, it's ancient.

>  What none of them do is what .NET does: return the argument.

Huh? Where/when does .NET do that?

.NET doesn't have an implementation of Cos/Sin of its own, it simply calls the relevant C library functions and those have gotten better other the time on all platforms. If those functions are bad, well, they need to be fixed but that's not something that .NET can do.

**Comment by dcwuser** [comment]
> Huh? Where/when does .NET do that?

Precisely where my bug report and the documentation (https://msdn.microsoft.com/en-us/library/system.math.sin(v=vs.110).aspx) say it does: for arguments greater than about 1.0E19. But thanks for illustrating that no one ever reads the documentation, making this bad behavior all the more dangerous. Please write a little .NET program that calls Math.Sin(1.0E22), and you will see that you get 1.0E22 back.

> I'm not sure how that's relevant, it's ancient.

And yet the vast majority of those ancient platforms had better behavior than .NET. (In my experience, nearly every new platform repeats the numerical mistakes that past platforms have long since fixed, because non-numerical programmers naively implement simple formulas without too much regard for numerical subtleties. I just got done fixing Complex.Sqrt to use the right algorithm, which was worked out in the 1960s.)

> .NET doesn't have an implementation of Cos/Sin of its own, it simply calls the relevant C library functions and those have gotten better other the time on all platforms.

This isn't quite true. If you include math.h into a native Win32 app and compute sin(1.0E22) and cos(1.0E22) you will get 0.412 and -0.911, which satisfies the -1 to 1 bounds and sin^2 + cos^2 = 1, so not exhibiting the bad .NET behavior and instead doing about the same thing as many of those "ancient", "irrelevant" systems from the 1990s. My understanding is that, presumably for portability, the .NET CLR includes its own native-level implementations of these functions, and that is where this bad behavior is.

This could be corrected either in .NET's native layer or by having the managed layer check bounds and throw ArgumentOutOfRangeException before calling into the native layer.

**Comment by mikedn** [comment]
> Precisely where my bug report and the documentation

Ah, that doc line. It was "copied" from the Intel documentation of FSIN/FCOS instructions that were used by the original .NET Framework. It doesn't accurately reflect the current .NET Core situation.

> Please write a little .NET program that calls Math.Sin(1.0E22), and you will see that you get 1.0E22 back.

I did an on .NET Core x64 I get 0.523214785395139. On x86 I get -0.91111900701809245, it's odd that it's different but it certainly isn't 1E22.

> My understanding is that, presumably for portability, the .NET CLR includes its own native-level implementations of these functions, and that is where this bad behavior is.

Not really. Some .NET versions/implementations simply use FSIN/FCOS instead of calling the C library functions. This made sense at the time because back when .NET was created the C compilers (well, VC++ at least) did pretty much the same thing. But this is not the case with .NET Core, see https://github.com/dotnet/coreclr/pull/4847 and https://github.com/dotnet/coreclr/pull/7023

**Comment by dcwuser** [comment]
> I did an on .NET Core x64 I get 0.523214785395139

Wow the platform-dependence here is more complicated than I thought.

So with a .NET Core 1.0 console app, compiled with Visual Studio 2015, with 
Microsoft .NET Core Tools (Preview 2) 14.1.21111.0 installed (which, I just checked, is still the latest release), on Windows 10 Pro 64-bit machine with Intel i7 CPU, I most definitely still get 1.0E22. That's also what I have got with all NETFX versions on all Windows machines I have tried this on since I became aware of this issue 4+ years ago.

But if I compile the .NET Core code from github source on that same machine, I also get 0.523214785395139 for Math.Cos(1.0E22). So is this new behavior since the preview 2 release? The dates on those merges appear to be before the preview 2 release date, but that doesn't necessarily mean they got into the release. Or will the new good behavior always morph into the old bad behavior when a production release is run on Windows? **Which is it?**

By the way, cos(1.0E22) = 0.523214785395139 is actually the right answer as computed by my own range-reduction logic.

**Comment by benaadams** [comment]
You likely need .NET Core 1.1 app to get the changes rather than 1.0

**Comment by mikedn** [comment]
> So is this new behavior since the preview 2 release?

Sorry, I'm not familiar with various past releases. What I know is that the issue doesn't exist in the master branch and so it won't exist in .NET Core 2.0 that is supposed to be released in a couple of months.

Note that this unfortunate behavior might persist in the desktop versions of .NET Framework, especially on x86. Sometimes changes made to .NET Core get merged back into .NET Framework but nobody knows when and if they'll do that.

>  Or will the new good behavior always morph into the old bad behavior when a production release is run on Windows?

Nope, that should not happen unless someone screws up the VC++ compiler options and get it to use the pesky FSIN/FCOS instructions instead of the C library implementation.

>  I most definitely still get 1.0E22

To sum it up:
* if you get 1e22 then the framework is using x87 FSIN/FCOS instructions
* if you get any other value (correct or not) the framework is using the C library functions
* the framework doesn't contain a sin/cos implementation of its own (unless you label the usage of FSIN/FCOS instructions as such)

**Comment by dcwuser** [comment]
I just verified @benaadams  assertion that, if you use the 1.1 runtime release with the 1.0.0-preview2-1-003177 SDK release, you get the new behavior.

This is absolutely fantastic and is, for my purposes, probably the single best change to come out of .NET core. It appears that the change also fixes more subtle bad behavior for lower arguments. (For example, the value at 104348/33215, a number close to \pi, used to be wrong in the last 4 digits, or about the last 35 bits, and those are now correct.) Almost certainly this means that the new library call is doing what I called the "read solution" above by performing range-reduction with a ~1100-bit accurate value of \pi. When this change hits NETFX, I will be able to pull out 100s of lines of my own code that was used to fix it.

Closing this bug. Thanks for the engagement and the fix.
